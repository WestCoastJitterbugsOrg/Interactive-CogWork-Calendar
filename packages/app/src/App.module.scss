.contentWrapper {
	display: flex;
	flex-flow: row wrap;
	align-items: stretch;
	background: #fff;
}

.eventSelection {
	display: flex;
	max-height: calc(100vh - 128px);
	width: 384px;
	flex-grow: 1;
	flex-direction: column;
}

.calendar {
	min-height: calc(100vh - 128px);
	min-width: max(calc(100% - 384px), 300px);
	flex-shrink: 0;
	flex-grow: 1;
}

.loader {
	width: 100%;
	position: relative;
	padding-top: 1px;
	background: var(--cw-color-secondary);

	.loaderBar {
		position: absolute;
		top: 0;
		bottom: 0;
		background: var(--cw-color-primary-alt);
		animation: borealis-bar 2s ease-in-out infinite;
	}

	@keyframes borealis-bar {
		0% {
			left: 0%;
			right: 100%;
			width: 0%;
		}
		10% {
			left: 0%;
			right: 75%;
			width: 25%;
		}
		90% {
			right: 0%;
			left: 75%;
			width: 25%;
		}
		100% {
			left: 100%;
			right: 0%;
			width: 0%;
		}
	}
}
