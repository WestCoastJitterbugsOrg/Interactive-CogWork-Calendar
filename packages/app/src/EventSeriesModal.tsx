import style from './EventSeriesModal.module.scss';
import closeIcon from './assets/close.svg';
import { useEffect, useRef } from 'react';
import Modal from 'react-modal';
import { Event } from './types/wcj';

type Props = {
	parent: HTMLElement;
	event: Event | undefined;
	close: () => void;
};

const rootHtmlElement = document.documentElement;

export function EventSeriesModal(props: Props) {
	const previousOverflowStyle = useRef<string>();
	useEffect(() => {
		// Code to prevent scrolling events getting triggered on the main page:
		if (props.event) {
			previousOverflowStyle.current = rootHtmlElement.style.overflow;
			rootHtmlElement.style.overflow = 'hidden';
		} else if (previousOverflowStyle.current) {
			// When modal is closed, we either have to reapply the style that was before
			rootHtmlElement.style.overflow = previousOverflowStyle.current;
		} else {
			// Or remove it if there was none
			rootHtmlElement.style.removeProperty('overflow');
		}
	}, [props.event]);

	return (
		<Modal
			onRequestClose={props.close}
			preventScroll={true}
			isOpen={!!props.event}
			className={style.modal}
			overlayClassName={style.overlay}
			parentSelector={() => props.parent}
			appElement={props.parent}
		>
			{props.event && (
				<div className={style.content}>
					<button
						type="button"
						className={style.closeButton}
						onClick={props.close}
						aria-label="modal-close-button"
					>
						<img alt="Close" src={closeIcon} width="16" height="16" />
					</button>
					<iframe
						className={style.iframe}
						src={props.event.registrationUrl}
						title={props.event.title}
					></iframe>
				</div>
			)}
		</Modal>
	);
}
