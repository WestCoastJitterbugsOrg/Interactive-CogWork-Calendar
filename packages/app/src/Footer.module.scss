.footer {
	display: flex;
	flex-wrap: wrap;
	place-items: center center;
	background-color: var(--cw-color-dark);
	padding: 8px;
}

.exportIcsButton {
	flex-grow: 1;
}

.exportIcsImg {
	height: 16px;
	width: 16px;
}

.githubWrapper {
	display: flex;
	flex-grow: 0;
	align-items: center;
	text-decoration-line: none;
}

.githubText {
	width: 100px;
	text-align: right;
	font-size: 12px;
	line-height: 16px;
	font-weight: 700;
	color: #fff;
}

.githubLogo {
	padding-left: 8px;
	padding-right: 8px;
}
