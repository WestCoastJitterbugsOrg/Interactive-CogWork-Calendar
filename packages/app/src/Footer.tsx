import style from './Footer.module.scss';
import fileDownload from './assets/file-download.svg';
import githubLogo from './assets/github-logo.png';
import { exportICS } from './services/ics';
import { Button } from './shared/Button';
import { stateContext } from './state';
import { useContext } from 'react';

export function Footer() {
	const { events, checkedEvents } = useContext(stateContext);

	return (
		<div className={style.footer}>
			<div className={style.exportIcsButton}>
				<Button
					onClick={() =>
						void exportICS(
							events.filter((event) => checkedEvents.includes(event.id)),
						)
					}
				>
					Exportera iCal{' '}
					<img
						alt="export icon"
						src={fileDownload}
						className={style.exportIcsImg}
					/>
				</Button>
			</div>
			<a
				className={style.githubWrapper}
				href="https://github.com/WestCoastJitterbugsOrg/Personalized-Calendar"
			>
				<span className={style.githubText}>
					Hj√§lp till med utveckling eller rapportera problem
				</span>
				<img
					alt="GitHub logo"
					src={githubLogo}
					className={style.githubLogo}
				></img>
			</a>
		</div>
	);
}
