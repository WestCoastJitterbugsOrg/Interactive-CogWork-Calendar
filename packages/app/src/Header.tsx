import style from './Header.module.scss';
import { getConsentCookie, storeConsentCookie } from './services/cookies';
import { Button } from './shared/Button';
import { useState } from 'react';

export function Header() {
	const [decisionMade, decide] = useState(false);

	const storeConsent = (consent: boolean) => {
		storeConsentCookie(consent ? 'yes' : 'no');
		decide(true);
	};

	if (getConsentCookie() || decisionMade) {
		return <></>;
	}
	return (
		<div className={style.header}>
			<div className={style.headerText}>
				<p>
					This calendar can remember which events are selected between sessions.
				</p>
				<p>We do not process this data for any other use.</p>
				<p>Your consent is needed for this functionality.</p>
			</div>
			<div style={{ display: 'flex' }}>
				<Button
					onClick={() => {
						storeConsent(true);
					}}
				>
					Consent
				</Button>
				<Button
					type="ghost"
					onClick={() => {
						storeConsent(false);
					}}
				>
					Don&apos;t consent
				</Button>
			</div>
		</div>
	);
}
