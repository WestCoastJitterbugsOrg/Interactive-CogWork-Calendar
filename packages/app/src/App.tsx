import appStyle from './App.module.scss';
import { EventSeriesModal } from './EventSeriesModal';
import { Footer } from './Footer';
import { Header } from './Header';
import { Calendar } from './calendar/Calendar';
import { EventSelection } from './event-selection/EventSelection';
import { stateContext } from './state';
import { useEffect, useState } from 'react';
import { WCJ } from './types';
import { canStoreSelection } from './services/cookies';

type Props = WCJ.Context & {
	colors: Record<string, string>;
	parent: HTMLElement;
	isLoading: boolean;
};

export function App(props: Props) {
	useEffect(() => {
		for (const color in props.colors) {
			const colorVal = props.colors[color];
			props.parent.style.setProperty(`--cw-color-${color}`, colorVal);
		}
	}, [props.colors, props.parent]);

	const [eventModal, setEventModal] = useState<string>();
	const [checkedEvents, setCheckedEvents] = useCheckedEvents(
		props.events.map((x) => x.id),
		props.selectedEventIds,
	);

	return (
		<stateContext.Provider
			value={{
				categories: props.categories,
				events: props.events,
				checkedEvents,
				setCheckedEvents,
				eventModal,
				setEventModal,
			}}
		>
			<Header />
			<div className={appStyle.contentWrapper}>
				<div className={appStyle.eventSelection} data-testid="event-selection">
					<EventSelection isLoading={props.isLoading} />
				</div>
				<div className={appStyle.calendar} data-testid="calendar">
					<Calendar events={props.events} checkedEvents={checkedEvents} />
				</div>
			</div>
			<Footer />
			<EventSeriesModal
				parent={props.parent}
				event={props.events.find((event) => event.id === eventModal)}
				close={() => {
					setEventModal(undefined);
				}}
			/>
		</stateContext.Provider>
	);
}

export function useCheckedEvents(
	allEventIds: string[],
	initialSelection: string[],
) {
	// This is a hook to store the user selection events into localstorage to keep it over multiple sessions.
	// Since new events can be added from one session to another, and we want the user to be able to see the new event,
	// we actually store the unselected events rather than the selected ones.

	const [checkedEvents, setCheckedEvents] =
		useState<string[]>(initialSelection);

	// When checkedEvents is changed, put it in localStorage
	useEffect(() => {
		if (canStoreSelection()) {
			const uncheckedEvents = allEventIds.filter(
				(eventId) => !checkedEvents.includes(eventId),
			);
			localStorage.setItem('uncheckedEvents', JSON.stringify(uncheckedEvents));
		}
	}, [allEventIds, checkedEvents]);

	return [checkedEvents, setCheckedEvents] as const;
}
