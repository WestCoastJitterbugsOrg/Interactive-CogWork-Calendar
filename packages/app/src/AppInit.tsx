import { App } from './App';
import style from './AppInit.module.scss';
import { initContext } from './services/cogwork';
import { ErrorViewer } from './shared/ErrorViewer';
import Loader from './shared/Loader';
import { useState } from 'react';
import type { WpCwfc } from 'shared/types';
import useSwr from 'swr';
import type { CW } from 'types';
import type { MaybeArray } from 'types/utils';

type Props = WpCwfc;

export default function AppInit(props: Props) {
	const formData = new FormData();
	formData.append('action', 'cwfc_fetch');
	formData.append('org', props.org);
	formData.append('pw_hash', props.pwHash);

	const [rootRef, setRef] = useState<HTMLElement | null>(null);

	const { isLoading, error, data } = useSwr<MaybeArray<CW.Event>, string>(
		'cwfc_fetch',
		() =>
			fetch(props.ajaxUrl, {
				method: 'POST',
				body: formData,
			}).then((res) =>
				res.ok ? res.json() : res.text().then((text) => Promise.reject(text))
			),
		{
			loadingTimeout: 10000,
		}
	);

	const getContent = () => {
		if (isLoading) {
			return <Loader />;
		}

		if (error) {
			return <ErrorViewer message={JSON.stringify(error, null, 4)} />;
		}

		if (!data) {
			return <h3>Didn&apos;t get any data</h3>;
		}

		if (rootRef)
			return (
				<App {...initContext(data)} colors={props.colors} parent={rootRef} />
			);
		return <></>;
	};

	return (
		<div className={style.root} ref={setRef}>
			{getContent()}
		</div>
	);
}
