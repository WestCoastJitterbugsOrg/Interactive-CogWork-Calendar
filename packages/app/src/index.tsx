import type { WpCwfc } from '../../shared/index';
import AppInit from './AppInit';
import { StrictMode } from 'react';
import { render as reactRender } from 'react-dom';
import { SWRConfig } from 'swr';

export function render(data: WpCwfc, container: Element) {
	return reactRender(
		<StrictMode>
			<SWRConfig value={{ provider: localStorageProvider }}>
				<AppInit {...data} />
			</SWRConfig>
		</StrictMode>,
		container
	);
}

function localStorageProvider() {
	// When initializing, we restore the data from `localStorage` into a map.
	const map = new Map(
		JSON.parse(localStorage.getItem('cwfc-cache') ?? '[]') as [string, object][]
	);

	// Before unloading the app, we write back all the data into `localStorage`.
	window.addEventListener('beforeunload', () => {
		const appCache = JSON.stringify(Array.from(map.entries()));
		localStorage.setItem('cwfc-cache', appCache);
	});

	// We still use the map for write & read for performance.
	return map;
}
