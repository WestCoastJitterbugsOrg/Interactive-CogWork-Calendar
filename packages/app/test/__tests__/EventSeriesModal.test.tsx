import '@testing-library/jest-dom';
import { RenderResult, fireEvent, render, act } from '@testing-library/react';
import { EventSeriesModal } from 'src/EventSeriesModal';
import AppMock from '../__mocks__/appMock';
import { mockStore } from '../__mocks__/stateContext';

it('EventSeriesModal Snapshot', () => {
	const { baseElement } = render(
		<EventSeriesModal
			parent={document.body}
			event={mockStore.events[0]}
			close={() => {}}
		/>,
	);

	expect(baseElement).toMatchSnapshot();
});

it('EventSeriesModal clicking on close button closes modal', async () => {
	const result = render(<AppMock />);
	await openModal(result);

	const modalCloseButton = await result.findByLabelText('modal-close-button');

	act(() => {
		fireEvent.click(modalCloseButton);
	});

	const modal =
		result.baseElement.querySelector<HTMLElement>('.ReactModalPortal');

	expect(modal).toBeEmptyDOMElement();
});

it('EventSeriesModal clicking escape closes modal', async () => {
	const result = render(<AppMock />);
	await openModal(result);
	act(() => {
		fireEvent.keyDown(
			result.baseElement.ownerDocument.activeElement ?? document.body,
			{
				keyCode: 27, // Escape
			},
		);
	});

	const modalPortal =
		result.baseElement.querySelector<HTMLElement>('.ReactModalPortal');
	expect(modalPortal).toBeEmptyDOMElement();
});

async function openModal(app: RenderResult) {
	const categoryElement = await app.findByText(mockStore.categories[0]);

	act(() => {
		// Click on the category to expand it
		fireEvent.click(categoryElement);
	});

	const infoButton = await app.findByAltText('info');
	act(() => {
		fireEvent.click(infoButton);
	});
}
