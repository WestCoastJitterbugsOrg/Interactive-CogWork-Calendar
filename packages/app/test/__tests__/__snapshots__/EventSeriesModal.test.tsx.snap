// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventSeriesModal Snapshot 1`] = `
<body
  aria-hidden="true"
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay"
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        aria-modal="true"
        class="ReactModal__Content"
        role="dialog"
        style="position: absolute; top: 40px; left: 40px; right: 40px; bottom: 40px; border: 1px solid #ccc; background: rgb(255, 255, 255); overflow: auto; border-radius: 4px; outline: none; padding: 20px;"
        tabindex="-1"
      >
        <div>
          <button
            aria-label="modal-close-button"
            type="button"
          >
            <img
              alt="Close"
              height="16"
              src="test-file-stub"
              width="16"
            />
          </button>
          <iframe
            src="https://example.com/registration"
            title="Grundkurs Lindy Hop"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
